File: src/Scripting/Ruby/patterns/data_mapper.rb
# frozen_string_literal: true

module RubyPatterns
  module DataMapper
    module_function

    def run
      table = { 1 => { name: 'Ada' } }
      user = Struct.new(:name)
      mapper = ->(row) { user.new(row[:name]) }
      raise 'Data Mapper failed' unless mapper.call(table[1]).name == 'Ada'
      true
    end
  end
end

RubyPatterns::DataMapper.run if $PROGRAM_NAME == __FILE__

File: src/Scripting/Lua/patterns/data_mapper.lua
local db = { [1] = { id=1, name="Ada" } }
local mapper = {}
function mapper.to_domain(row) return { id=row.id, display_name=row.name } end
local person = mapper.to_domain(db[1])
assert(person.display_name == "Ada" and person.name == nil)
return true
